### ISSUE TITLE
MPRIS integration

### ISSUE THREAD

**Issue by rosshadden** [cfangmeier/tuijam#12]
Do you have any plans to make tuijam work with the MPRIS D-Bus spec, which would let us use [playerctl](https://github.com/acrisci/playerctl) and similar programs?

Not only would this let users control playback externally, but it would allow for showing track, artist, state, etc. externally as well (such as in polybar or i3bar).

**Comment by cfangmeier** [comment]
I didn't have plans for this (nor did I know anything about MPRIS), but it sounds like a very useful feature!

I found a python wrapper for D-Bus, `pydbus`, which should reduce the problem to just rigging up the appropriate D-Bus paths to call a few functions in TUIJam.

I have a bit of time today so I'll give it a shot and see if I can get it working with playerctl and push what I have into a dev branch.

Would you be willing to test out the result on your machine?

**Comment by rosshadden** [comment]
Yeah of course. Thank you.

**Comment by cfangmeier** [comment]
Alright, I believe I've got it working. Checkout the `mpris` branch to try it out. On my machine, I am able to query the status, set play/pause, and get some metadata for the current song using `playerctl`. For example,

```bash
caleb@narsil ~
> $ playerctl -p tuijam metadata
{'mpris:trackid': <'Tbgjybvtdkktneyurpnpztidcby'>, 'xesam:title': <'Salt + Charcoal'>, 'xesam:artist': <'Plini'>, 'xesam:album': <'Sunhead'>}
```

Let me know what does/doesn't work for you and if there are any missing features you would like to see.

**Comment by rosshadden** [comment]
It works!

I did found a couple weird things:

1. Switching to the next track works, but previous doesn't
2. Pausing works, but playing afterward doesn't (not sure if that's in your control, and it also doesn't matter to me since play-pause works great)
3. Same thing with stopping. It works, but playing afterward doesn't start the song over. It errors instead.

But everything you mentioned is working as well. Actually on a related not to 1., I just realized there is no default shortcut for previous track. I'll probably make a separate issue for this and maybe add it myself if I get a chance.

**Comment by rosshadden** [comment]
Thank you so much for working on this, by the way. I have tried many Google Music CLI clients. My favorite workflow was mopidy+gmusicproxy+ncmpcpp, but it stopped working after a year of using it, and despite many hours and days of troubleshooting (often while bored at work) I never figured it out. This project is definitely far better than any others I have tried (with the exception of the aforementioned now nonworking one), and I am excited to be able to play Google Music through the terminal again.

And thanks for your quick response and work on the MPRIS stuff.
